src/ast/dibuilderbpf.cpp

```diff
@@ -59,14 +86,182 @@ DIType *DIBuilderBPF::getInt64Ty()
+    types_.int8_ptr = createPointerType(getInt8Ty(), 64);
 
   return types_.int8_ptr;
 }
 
+// Create anonymous struct with anonymous fields. It's possible that there will
+// be multiple tuples of the same (duplicated) type but BTF deduplication should
+// take care of that.
+DIType *DIBuilderBPF::CreateTupleType(const SizedType &stype)
+{
+  assert(stype.IsTupleTy());
+
+  SmallVector<Metadata *, 8> fields;
+  for (auto &field : stype.GetFields())
+  {
+    fields.push_back(createMemberType(file,
+                                      "",
+                                      file,
+                                      0,
+                                      field.type.GetSize() * 8,
+                                      0,
+                                      field.offset * 8,
+                                      DINode::FlagZero,
+                                      GetType(field.type)));
+  }
+  DICompositeType *result = createStructType(file,
+                                             "",
+                                             file,
+                                             0,
+                                             stype.GetSize() * 8,
+                                             8,
+                                             DINode::FlagZero,
+                                             nullptr,
+                                             getOrCreateArray(fields));
+  return result;
+}
+
+DIType *DIBuilderBPF::GetType(const SizedType &stype)
+{
+  if (stype.IsByteArray() || stype.IsRecordTy())
+  {
+    auto subrange = getOrCreateSubrange(0, stype.GetSize());
+    return createArrayType(
+        stype.GetSize() * 8, 0, getInt8Ty(), getOrCreateArray({ subrange }));
+  }
+
+  if (stype.IsArrayTy())
+  {
+    auto subrange = getOrCreateSubrange(0, stype.GetNumElements());
+    return createArrayType(stype.GetSize() * 8,
+                           0,
+                           GetType(*stype.GetElementTy()),
+                           getOrCreateArray({ subrange }));
+  }
+
+  if (stype.IsTupleTy())
+    return CreateTupleType(stype);
+
+  if (stype.IsPtrTy())
+    return getInt64Ty();
+
+  // Integer types and builtin types represented by integers
+  switch (stype.GetSize())
+  {
+    case 8:
+      return getInt64Ty();
+    case 4:
+      return getInt32Ty();
+    case 2:
+      return getInt16Ty();
+    case 1:
+      return getInt8Ty();
+    default:
+      LOG(FATAL) << "Cannot generate debug info for type "
+                 << typestr(stype.type) << " (" << stype.GetSize()
+                 << " is not a valid type size)";
+      return nullptr;
+  }
+}
+
+DIType *DIBuilderBPF::GetMapKeyType(const MapKey &key,
+                                    const SizedType &value_type)
+{
+  // No-key maps use '0' as the key.
+  // No-key count() maps use BPF_MAP_TYPE_PERCPU_ARRAY which needs 4-byte key.
+  if (key.args_.size() == 0)
+    return value_type.IsCountTy() ? getInt32Ty() : getInt64Ty();
```

Comment:
Just curious, why do the other no-key maps need a 64 int?

Reply:
Normally, we use 64-bit ints everywhere, just BPF_MAP_TYPE_PERCPU_ARRAY needs 32-bit keys [per documentation](https://docs.kernel.org/bpf/map_array.html).